File: app/src/main/java/me/zhanghai/android/files/fileproperties/permissions/SetPrincipalViewModel.java
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.fileproperties.permissions;

import android.text.TextUtils;

import java.util.Objects;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MediatorLiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import me.zhanghai.android.files.util.SelectionLiveData;
import me.zhanghai.java.functional.Functional;

public abstract class SetPrincipalViewModel extends ViewModel {

    @NonNull
    private final MutableLiveData<PrincipalListData> mPrincipalListLiveData =
            createPrincipalListLiveData();

    @NonNull
    private final MutableLiveData<String> mFilterLiveData = new MutableLiveData<>();

    @NonNull
    private final LiveData<PrincipalListData> mFilteredPrincipalListLiveData =
            new FilteredPrincipalListLiveData(mPrincipalListLiveData, mFilterLiveData);

    @NonNull
    private final SelectionLiveData<Integer> mSelectionLiveData = new SelectionLiveData<>();

    protected abstract MutableLiveData<PrincipalListData> createPrincipalListLiveData();

    public void setFilter(@NonNull String filter) {
        if (!Objects.equals(mFilterLiveData.getValue(), filter)) {
            mFilterLiveData.setValue(filter);
        }
    }

    @NonNull
    public PrincipalListData getPrincipalListData() {
        return mPrincipalListLiveData.getValue();
    }

    @NonNull
    public LiveData<PrincipalListData> getFilteredPrincipalListLiveData() {
        return mFilteredPrincipalListLiveData;
    }

    @NonNull
    public SelectionLiveData<Integer> getSelectionLiveData() {
        return mSelectionLiveData;
    }

    private static class FilteredPrincipalListLiveData extends MediatorLiveData<PrincipalListData> {

        @NonNull
        private LiveData<PrincipalListData> mPrincipalListLiveData;
        @NonNull
        private LiveData<String> mFilterLiveData;

        public FilteredPrincipalListLiveData(
                @NonNull LiveData<PrincipalListData> principalListLiveData,
                @NonNull LiveData<String> filterLiveData) {

            mPrincipalListLiveData = principalListLiveData;
            mFilterLiveData = filterLiveData;

            addSource(mPrincipalListLiveData, userList -> loadValue());
            addSource(mFilterLiveData, filter -> loadValue());
        }

        private void loadValue() {
            String filter = mFilterLiveData.getValue();
            PrincipalListData principalListData = mPrincipalListLiveData.getValue();
            if (!TextUtils.isEmpty(filter)) {
                principalListData = principalListData.filter(user -> filterPrincipal(user, filter));
            }
            setValue(principalListData);
        }

        private static boolean filterPrincipal(@NonNull PrincipalItem principal,
                                               @NonNull String filter) {
            return Integer.toString(principal.id).contains(filter)
                    || (principal.name != null && principal.name.contains(filter))
                    || (Functional.some(principal.applicationInfos, applicationInfo ->
                    applicationInfo.packageName.contains(filter)))
                    || (Functional.some(principal.applicationLabels, applicationLabel ->
                    applicationLabel.contains(filter)));
        }
    }
}

File: app/src/main/java/me/zhanghai/android/files/fileproperties/permissions/SetPrincipalViewModel.kt
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.fileproperties.permissions

import androidx.lifecycle.LiveData
import androidx.lifecycle.MediatorLiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import me.zhanghai.android.files.util.SelectionLiveData
import me.zhanghai.android.files.util.valueCompat

abstract class SetPrincipalViewModel(
    private val principalListLiveData: MutableLiveData<PrincipalListData>
) : ViewModel() {
    val principalListData: PrincipalListData
        get() = principalListLiveData.valueCompat

    private val filterLiveData = MutableLiveData("")
    var filter: String
        get() = filterLiveData.valueCompat
        set(value) {
            if (filterLiveData.valueCompat != value) {
                filterLiveData.value = value
            }
        }

    val filteredPrincipalListLiveData: LiveData<PrincipalListData> =
        FilteredPrincipalListLiveData(principalListLiveData, filterLiveData)

    val selectionLiveData = SelectionLiveData<Int>()

    private class FilteredPrincipalListLiveData(
        private val principalListLiveData: LiveData<PrincipalListData>,
        private val filterLiveData: LiveData<String>
    ) : MediatorLiveData<PrincipalListData>() {
        init {
            addSource(principalListLiveData) { loadValue() }
            addSource(filterLiveData) { loadValue() }
        }

        private fun loadValue() {
            val filter = filterLiveData.valueCompat
            var principalListData = principalListLiveData.valueCompat
            if (filter.isNotEmpty()) {
                principalListData = principalListData.filter { filterPrincipal(it, filter) }
            }
            value = principalListData
        }

        private fun filterPrincipal(principal: PrincipalItem, filter: String): Boolean =
            (filter in principal.id.toString()
                || (principal.name != null && filter in principal.name)
                || principal.applicationInfos.any { filter in it.packageName }
                || principal.applicationLabels.any { filter in it })
    }
}
